core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -510,33 +489,28 @@ private void cleanFailedShards(final ClusterChangedEvent event) {
             return;
         }
         DiscoveryNodes nodes = event.state().nodes();
-        long now = System.currentTimeMillis();
         String localNodeId = nodes.localNodeId();
-        Iterator<Map.Entry<ShardId, FailedShard>> iterator = failedShards.entrySet().iterator();
-        shards:
+        Iterator<Map.Entry<ShardId, ShardRouting>> iterator = failedShards.entrySet().iterator();
+
         while (iterator.hasNext()) {
-            Map.Entry<ShardId, FailedShard> entry = iterator.next();
-            FailedShard failedShard = entry.getValue();
-            IndexRoutingTable indexRoutingTable = routingTable.index(entry.getKey().getIndex());
+            Map.Entry<ShardId, ShardRouting> entry = iterator.next();
+            ShardId failedShardId = entry.getKey();
+            ShardRouting failedShardRouting = entry.getValue();
+            IndexRoutingTable indexRoutingTable = routingTable.index(failedShardId.getIndex());
             if (indexRoutingTable != null) {
-                IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(entry.getKey().id());
+                IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(failedShardId.id());
                 if (shardRoutingTable != null) {
                     for (ShardRouting shardRouting : shardRoutingTable.assignedShards()) {
                         if (localNodeId.equals(shardRouting.currentNodeId())) {
```

Comment:
I don't think we need the filtering on the node. How about using:

```
shardRoutingTable.assignedShards().stream().filter(as -> as.isSameAllocation(failedShard)).findFirst().isPresent()
```

If true, we keep the shard failure, o.w. we clear the iterator...